azerothcore/azerothcore-wotlk
# fix(DB/spell): Adjust Intercept and Charge attrs

Intercept and Charge stun and damage effects were dodgeable.

The attributes of Intercept (damage & stun) (id: 20253) and Charge
(stun) (id: 7922) have been adjusted to include a flag that makes
these effects impossible to dodge, parry and block.



## Changes Proposed:
- add SPELL_ATTR0_IMPOSSIBLE_DODGE_PARRY_BLOCK flag to the aforementioned spells
(it's called IMPOSSIBLE_DODGE_PARRY_BLOCK in the docs (https://wowdev.wiki/Spell.dbc/Attributes) and NO_ACTIVE_DEFENSE internally)

## Issues Addressed:

- Closes #6400
- Closes chromiecraft/chromiecraft#864

## SOURCE:

Refer to chromiecraft/chromiecraft#864

## Tests Performed:

- Sql applies cleanly
- Tested in-game. Testing one of three cases (dodge/parry/block) should be sufficient because they're all covered by the flag (look for `SPELL_ATTR0_NO_ACTIVE_DEFENSE` in `Unit.cpp` if you want to make sure)

## How to Test the Changes:


I'm sure there's a less cumbersome way of testing this. I used the following steps:
1. Buff the Greater Agility elixir to grant 2000 (+1) agility. Use the included statement: https://pastebin.com/zQt9Y334
2. Create a level 40-ish character, drink elixir which will increase dodge to 80-ish %
3. Create a level 40-ish character of the opposite faction, `.learn` Charge, Bloodrage, Intercept (100, 2687, 20252)
4. Set pvp on on both characters
5. Intercept the high dodge character a couple times, make sure the yellow damage never gets dodged and the target gets stunned
6. Charge the high dodge character a couple times, make sure the target always get stunned

Use Bloodrage to generate rage for Intercept, use `.cheat cooldown` to save time, use a `.gm on .gm off` macro to get out of combat inbetween Charge casts.

Note that after getting stunned three times in quick succession, targets gain stun immunity for a short period. This means that Intercepted targets do not get stunned (but they still take damage), and Charged targets will be "Immune" to the stun.


## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
